grub-core/loader/multiboot_elfxx.c

```diff
@@ -180,6 +259,26 @@ CONCAT(grub_multiboot_load_elf, XX) (mbi_load_data_t *mld)
         }
     }
 
+  if (grub_slaunch_platform_type () == SLP_INTEL_TXT)
+    {
+      slparams->mle_header_offset = 0xffffffff;
+
+      for (mle_hdr_offset = 0;  mle_hdr_offset < 0x1000; mle_hdr_offset += 16)
+	{
+	  if ( !grub_memcmp ((void *)(slparams->mle_start + mle_hdr_offset), GRUB_TXT_MLE_UUID, 16) )
```

Comment:
`slparams->mle_start` is `physical_target_address`, you shouldn't assume it is already populated. You must check against `virtual_current_address`, i.e. at `source`, even if in this case they are the same.

Reply:
`slparams->mle_start` is already populated ~100 lines of code before. Secondly, `source` may change when loadable segments are loaded to memory, so I can't rely on that.

Reply:
The address is populated, but if data needs to be relocated, it is done by relocator just before handing off execution to payload. You can also move these lines closer to where `source` is saved.

Reply:
Changed to source